objax/optimizer/lars.py

```diff
@@ -0,0 +1,78 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+__all__ = ['LARS']
+
+from typing import List
+
+import jax.numpy as jn
+
+from objax.module import Module, ModuleList
+from objax.typing import JaxArray
+from objax.variable import TrainRef, StateVar, TrainVar, VarCollection
+
+
+class LARS(Module):
+    """Layerwise adaptive rate scaling (LARS) optimizer.
+
+    See https://arxiv.org/abs/1708.03888
+    """
+
+    def __init__(self, vc: VarCollection,
+                 momentum: float = 0.9,
+                 beta: float = 1e-4,
+                 tc: float = 1e-3,
+                 eps: float = 1e-5,
+                 max_epochs: int = 100):
+        """Constructor for LARS optimizer.
+
+        Args:
+            vc: collection of variables to optimize.
+            momentum: coefficient used for the moving average of the gradient.
+            beta: weight decay coefficient.
+            tc: trust coefficient eta ( < 1) for trust ratio computation.
+            eps: epsilon used for trust ratio computation.
+            max_epochs: total number of steps that the optimizer will take.
```

Comment:
If I understand well what is called here `max_epoch` is the parameter `T` in the paper.
From the look of it, it's a built-in learning rate decay which should be independent of the method itself for the sake of generality.

What I mean by that is that one could simply invoke `__call__` with `((1-t/T)**2, grads)` and not be tied to a polynomial learning rate decay.

Reply:
OK. I removed the LR decay logic and updated the doc string to indicate that the paper suggested calling the optimizer with a decaying LR.